LP-1782 Implement table widget on LMS side for both the reviewer and the attempter

qasimgulzar opened the pull request.

Commit:
[PARTIAL] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter
[FIX] "Save Progress" button should be active, If user just entered responses in ‘Table’.

Commit:
[PARTIAL] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter
[FIX] provide support for firefox.

Commit:
[PARTIAL] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter
[FIX] restrict user to add at least one prompt to open assessment.
[REMOVE] `console.log`
[FIX] if there is only one prompt in assessment it is renders as default.

Commit:
[COMPLETE] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter
[FIX] provide support to edit the existing table.

Commit:
[COMPLETE] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter
[FIX] populate peer assessment checkboxes.

emraan004 commented on openassessment/templates/openassessmentblock/edit/oa_edit_prompt.html:
```diff
@@ -20,12 +20,25 @@ <h2>{% trans "Remove" %}</h2>
-            <li class="field comp-setting-entry openassessment_prompt_description_wrapper">
+            <li class="field comp-setting-entry openassessment_prompt_description_wrapper enable-header-editing">
                 {{html_content|safe}}
+                <script type="text/javascript">
```
Can we move this inline method to a separate file? Also let's use constants for selectors

emraan004 commented on openassessment/templates/openassessmentblock/edit/oa_edit_table_prompt.html:
```diff
@@ -16,7 +16,11 @@
-        <p class="text-input" contenteditable="true"></p>
+        <div class="text-input" data-input-control="true" data-input-control-type="textarea" data-hidden-place-holder=".oa-html-content-text-place-holder">
+            <textarea
+                    onkeyup="(function (element) {$(element).parent().find('p').first().text($(element).val());})(this)"/>
```
please move this inline function to a separate file and make it readable

emraan004 commented on openassessment/templates/openassessmentblock/oa_submission_answer.html:
```diff
@@ -13,14 +13,14 @@ <h5 class="submission__answer__part__text__title">{% trans "The question for thi
                         {{ part.text|safe }}
                     </div>
                     <script type="text/javascript">
                         $(document).ready(function () {
```
can we discuss the reason behind writing this inline JS functions?

emraan004 commented on openassessment/templates/openassessmentblock/response/oa_response.html:
```diff
@@ -63,13 +63,26 @@ <h5 class="submission__answer__part__text__title">{% trans "Prompt for this sect
                                                     $(oa_block_saved_id).find('td').each(function (index, element) {
-                                                        $(oa_table_td_list.get(index)).html($(element).html());
-                                                        $(oa_table_td_list.get(index)).find('.text-input').attr({contenteditable: true});
+                                                        var text_input=$(oa_table_td_list.get(index)).find('.text-input');
```
same here inline function

ahmed-faraz46 replied:
Let's, please make sure that this line is properly indented.

emraan004 commented on openassessment/xblock/static/js/openassessment-lms.min.js:
```diff
@@ -1,3 +1,3 @@
-$.fn.oraTableBuilder=function($config){var CHECKBOX_SELECTOR='input[type="checkbox"]',CONTENTENDITABLE_SELECTOR="[contenteditable]";var _this=this;var change_cb=function(table){console.info("Please overrider change in configuration object")};if($config&&$config.change){change_cb=$config.change}else{$config.change=change_cb}function plugin(){this._this=_this;var __this=this;this.controls={div:$("< ...
-installHandlers:function(){this.baseView.setUpCollapseExpand($(".step--leaderboard",this.element))}};OpenAssessment.MessageView=function(element,server,baseView){this.element=element;this.server=server;this.baseView=baseView};OpenAssessment.MessageView.prototype={load:function(){var view=this;var baseView=this.baseView;this.server.render("message").done(function(html){$(".openassessment__message" ...
-}).fail(function(){view.baseView.showLoadError("student-training")})},installHandlers:function(){var sel=$(".step--student-training",this.element);var view=this;this.baseView.setUpCollapseExpand(sel);var rubricSelector=$(".student-training--001__assessment",this.element);if(rubricSelector.size()>0){var rubricElement=rubricSelector.get(0);this.rubric=new OpenAssessment.Rubric(rubricElement)}if(thi ...
```
why is this min file getting pushed? doesn't our deployment generate new ones?

qasimgulzar replied:
as per my understanding as going through the code my findings are that I would not compile static files of external packages.

emraan004 commented on openassessment/xblock/static/js/philu/lib/table_builder.js:
```diff
@@ -359,9 +381,53 @@ $.fn.oaTable = function () {
+    }
+
+    return new plugin();
+}
```
EOP

Commit:
[COMPLETE] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter
[FIX] edit table header textarea populated with garbage.

Commit:
[COMPLETE] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter.
[CHANGE] move inline js code to separate file.

Commit:
[COMPLETE] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter.
[CHANGE] add end of file.

Commit:
[COMPLETE] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter.
[FIX] editing table headers converts it to default prompt.

ahmed-faraz46 commented on Makefile:
```diff
@@ -31,7 +31,7 @@ STATIC_JS = openassessment/xblock/static/js
 javascript:
 	node_modules/.bin/uglifyjs $(STATIC_JS)/philu/lib/table_builder.js $(STATIC_JS)/src/oa_shared.js $(STATIC_JS)/src/*.js $(STATIC_JS)/src/lms/*.js > "$(STATIC_JS)/openassessment-lms.min.js"
-	node_modules/.bin/uglifyjs $(STATIC_JS)/philu/lib/table_builder.js $(STATIC_JS)/src/oa_shared.js $(STATIC_JS)/src/*.js $(STATIC_JS)/src/studio/*.js $(STATIC_JS)/philu/studio/*.js > "$(STATIC_JS)/openassessment-studio.min.js"
+	node_modules/.bin/uglifyjs $(STATIC_JS)/philu/lib/helpers.js $(STATIC_JS)/philu/lib/table_builder.js $(STATIC_JS)/src/oa_shared.js $(STATIC_JS)/src/*.js $(STATIC_JS)/src/studio/*.js $(STATIC_JS)/philu/studio/*.js > "$(STATIC_JS)/openassessment-studio.min.js"
```
How about `$(STATIC_JS)/philu/lib/*`?

qasimgulzar replied:
can't add * to ensure the order, in new commit helpers will be moved to end.

ahmed-faraz46 commented on openassessment/templates/openassessmentblock/edit/partial/widget_templates.html:
```diff
@@ -0,0 +1,30 @@
+<script type="text/html" data-w-name="input" data-w-type="checkbox">
```
Would you please mention why do we need to wrap these elements in a script? Why can't we pass the data attr directly to <label>?

qasimgulzar replied:
it is wrapped just to clone later.

ahmed-faraz46 commented on openassessment/templates/openassessmentblock/edit/partial/widget_templates.html:
```diff
@@ -0,0 +1,30 @@
+            Horizontal Headers
+        </label>
+    </div>
+</script>
```
Let's please add EOF.

monofal replied:
lets add eof here.

ahmed-faraz46 commented on openassessment/templates/openassessmentblock/edit/oa_edit_prompt.html:
```diff
@@ -20,12 +20,25 @@ <h2>{% trans "Remove" %}</h2>
+                    $(document).ready(function () {
+                        var config = {
+                            update: function (table) {
+                                var targerTextArea = $('#html-content-textarea-{{ counter }}'),
```
Let's please fix this typo, `targer`

ahmed-faraz46 commented on openassessment/templates/openassessmentblock/edit/oa_edit_prompt.html:
```diff
@@ -20,12 +20,25 @@ <h2>{% trans "Remove" %}</h2>
+                                targerTextArea.val(updated_html);
+                            }
+                        };
+                        $('.enable-header-editing').openassessmentEditableTableHeaders(config);
```
How about to use id in this selector as well, because if there are multiple `html-control` widgets in a single ORA this selector will be for all of them, but here inside the loop you're want to call `openassessmentEditableTableHeaders` for a specific li.

qasimgulzar replied:
we have handled it in other way

ahmed-faraz46 commented on openassessment/templates/openassessmentblock/edit/oa_edit_prompt.html:
```diff
@@ -20,12 +20,25 @@ <h2>{% trans "Remove" %}</h2>
+                            update: function (table) {
+                                var targerTextArea = $('#html-content-textarea-{{ counter }}'),
+                                    updated_html = table.html();
+                                table.find('script').remove();
```
Why did we add <script> if we need to remove them?

qasimgulzar replied:
to make lockable html.

monofal commented on openassessment/xblock/static/js/philu/lib/helpers.js:
```diff
@@ -0,0 +1,4 @@
+$(document).on("keyup",".text-input textarea", function(){
+    var element = $(this);
+    $(element).next().text($(element).val());
+});
```
lets add eof here.

ahmed-faraz46 replied:
EOF please.

ahmed-faraz46 commented on openassessment/templates/openassessmentblock/edit/oa_edit_prompt.html:
```diff
@@ -20,12 +20,25 @@ <h2>{% trans "Remove" %}</h2>
             <li class="field comp-setting-entry openassessment_prompt_description_wrapper">
```
Let's please try to move core changes somewhere else and use that file here so ut will result in just one core change.
Second way is to add a seprate `file oa_edit_table_prompt` and use that file in `oa_edit_prompts.html` so in that file we will be checking widget type and rendering respective editor. And I see this file is already part of the code

qasimgulzar replied:
@ahmed-faraz46 we have moved most of the code changes to separate file.

monofal commented on openassessment/xblock/static/js/philu/studio/oa_container_item_overriders.js:
```diff
@@ -24,15 +23,13 @@ OpenAssessment.Prompt.prototype.getFieldValues = function () {
 OpenAssessment.Prompt.prototype.getHtmlContent = function () {
-    console.log('OpenAssessment.Prompt.prototype.getHtmlContent')
     return $('.openassessment_prompt_html_content', this.element).first().val();
 }
```
lets add eof here.

ahmed-faraz46 replied:
EOF please!

ahmed-faraz46 commented on openassessment/templates/openassessmentblock/edit/oa_edit_prompts.html:
```diff
@@ -16,7 +16,7 @@
     <ul id="openassessment_prompts_list">
         {% for prompt in prompts %}
-            {% include "openassessmentblock/edit/oa_edit_prompt.html" with prompt_uuid=prompt.uuid prompt_description=prompt.description w_type=prompt.prompt_type html_content=prompt.html_content %}
+            {% include "openassessmentblock/edit/oa_edit_prompt.html" with prompt_uuid=prompt.uuid prompt_description=prompt.description w_type=prompt.prompt_type html_content=prompt.html_content counter=forloop.counter %}
```
If we taking the approach to make `oa_edit_table_prompt` then we will implement sanity here to include corresponding file

ahmed-faraz46 commented on openassessment/xblock/static/js/philu/lib/helpers.js:
```diff
@@ -0,0 +1,4 @@
+$(document).on("keyup",".text-input textarea", function(){
+    var element = $(this);
```
we can receive `event` in this callback and then we can use event.target directly instead making element from this and then using that element.

qasimgulzar replied:
we are not making elements here it is just to convert native dom element to jquery object.

ahmed-faraz46 commented on openassessment/xblock/validation.py:
```diff
@@ -373,6 +373,9 @@ def _inner(rubric_dict, assessments, leaderboard_show=0, submission_start=None,
         submission_dates = [(submission_start, submission_due)]
         assessment_dates = [(asmnt.get('start'), asmnt.get('due')) for asmnt in assessments]
         success, msg = validate_dates(oa_block.start, oa_block.due, submission_dates + assessment_dates, _)
+        if not success:
```
Let's please add a wiki in ORA document all core changes in ORA, please follow the template which is mentioned in https://github.com/philanthropy-u/edx-platform/wiki/Core-Changes

## Title: This is the title of core change
This is a description of the core change
##### Why this core change is inevitable?
The reason why this core change is inevitable.
##### Why you can't move this change to `overrides`?
The reason why this core change can't be moved to overrides.
##### Link of the ticket?
[LP-XXXX](https://philanthropyu.atlassian.net/browse/LP-XXXX)
##### Link of the PR?
[PR](https://philanthropyu.atlassian.net/browse/LP-XXXX)
*****

Commit:
[COMPLETE] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter.
[REFACTORING] cleaning javascript code and moving it to external file.

Commit:
[COMPLETE] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter.
[REMOVE] unused code.

emraan004 commented on openassessment/xblock/static/js/philu/lib/helpers.js:
```diff
@@ -0,0 +1,82 @@
+            }
+
+            event.preventDefault();
+            $(this).find('li.submission__answer__part[data-prompt-type="html-control"]').each(function (index, oa_block_li) {
```
let's make constants for all dom element selectors

qasimgulzar replied:
done

Commit:
[COMPLETE] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter.
[REFACTORING] make selector constants.

Commit:
[COMPLETE] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter.
[FIX] remove html tags out of text while editing headers.
[FIX] make prompt description style consistent.

Commit:
[COMPLETE] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter.
[UPDATE] updated minified javascript files.

Commit:
[COMPLETE] LP-1782 Implement table widget on LMS side for both the reviewer and the attempter.
[CHANGE] add docstrings to function.

monofal reviewed (approved):
LGTM 👍

emraan004 reviewed (approved):
👍
